pallets/pallet-bonded-coins/src/curves/lmsr.rs

```diff
@@ -49,48 +66,25 @@ where
 	Parameter: FixedSigned + PartialOrd<Precision> + From<Precision> + ToFixed,
 	<Parameter as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign,
 {
-	// c(a, c) = (a - c) + b * ln((1 + SUM_i e^((q_i - a)/b)) / (1 + SUM_i e^((q_i - c)/b)))
 	fn calculate_costs(
 		&self,
 		low: Parameter,
 		high: Parameter,
 		passive_supply: PassiveSupply<Parameter>,
 	) -> Result<Parameter, ArithmeticError> {
-		let e_term_numerator = passive_supply
-			.iter()
-			.map(|x| self.calculate_exp_term(*x, high))
-			.collect::<Result<Vec<Parameter>, ArithmeticError>>()?;
-
-		let term1 = e_term_numerator.iter().try_fold(Parameter::from_num(0), |acc, x| {
-			acc.checked_add(*x).ok_or(ArithmeticError::Overflow)
-		})?;
-
-		let numerator = Parameter::from_num(1)
-			.checked_add(term1)
-			.ok_or(ArithmeticError::Overflow)?;
-
-		let e_term_denominator = passive_supply
-			.iter()
-			.map(|x| self.calculate_exp_term(*x, low))
-			.collect::<Result<Vec<Parameter>, ArithmeticError>>()?;
-
-		let term2 = e_term_denominator.iter().try_fold(Parameter::from_num(0), |acc, x| {
-			acc.checked_add(*x).ok_or(ArithmeticError::Overflow)
-		})?;
-
-		let denominator = Parameter::from_num(1)
-			.checked_add(term2)
-			.ok_or(ArithmeticError::Overflow)?;
-
-		let log_value = numerator
-			.checked_div(denominator)
-			.ok_or(ArithmeticError::DivisionByZero)
-			.and_then(|x| ln::<Parameter, Parameter>(x).map_err(|_| ArithmeticError::Overflow))?;
-
-		let high_low_diff = high.checked_sub(low).ok_or(ArithmeticError::Underflow)?;
+		// Clone passive_supply and add low and high to create modified supplies
+		let mut low_total_supply = passive_supply.clone();
+		low_total_supply.push(low);
+		let mut high_total_supply = passive_supply.clone();
+		high_total_supply.push(high);
 
-		let m_log_value = self.m.checked_mul(log_value).ok_or(ArithmeticError::Overflow)?;
+		// Compute LSE for both modified supplies
+		let lower_bound_value = self.lse(&low_total_supply)?;
```

Comment:
looks like we would expect to be able to compute the lse of an all-0 supply vector. In that case we need to fall back to 0 where we would try to compute the logarithm of 0, correct?

Reply:
see https://github.com/KILTprotocol/kilt-node/pull/777#discussion_r1827767284